components/ai_chat/core/browser/conversation_driver.cc

```diff
@@ -310,106 +312,160 @@ bool ConversationDriver::MaybePopPendingRequests() {
+  // an in-progress operation.
+  if (is_page_text_fetch_in_progress_) {
+    VLOG(1) << "A page content fetch is in progress, waiting for the existing "
+               "operation to complete";
+    auto handle_existing_fetch_complete = base::BindOnce(
+        &ConversationDriver::OnExistingGeneratePageContentComplete,
+        weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+    on_page_text_fetch_complete_->Post(
+        FROM_HERE, std::move(handle_existing_fetch_complete));
     return;
   }
 
-  if (!HasPrimaryMainFrame()) {
-    VLOG(1) << "Summary request submitted for a WebContents without a "
-               "primary main frame";
-    return;
-  }
+  is_page_text_fetch_in_progress_ = true;
+  // Update fetching status
+  OnPageHasContentChanged(BuildSiteInfo());
 
-  if (should_send_page_contents_) {
-    is_page_text_fetch_in_progress_ = true;
-    // Update fetching status
-    OnPageHasContentChanged(BuildSiteInfo());
-    GetPageContent(
-        base::BindOnce(&ConversationDriver::OnPageContentRetrieved,
-                       weak_ptr_factory_.GetWeakPtr(), current_navigation_id_));
-  }
+  GetPageContent(
+      base::BindOnce(&ConversationDriver::OnGeneratePageContentComplete,
+                     weak_ptr_factory_.GetWeakPtr(), current_navigation_id_,
+                     std::move(callback)),
+      content_invalidation_token_);
 }
 
-void ConversationDriver::OnPageContentRetrieved(int64_t navigation_id,
-                                          std::string contents_text,
-                                          bool is_video) {
+void ConversationDriver::OnGeneratePageContentComplete(
+    int64_t navigation_id,
+    GetPageContentCallback callback,
+    std::string contents_text,
+    bool is_video,
+    std::string invalidation_token) {
+  VLOG(1) << "OnGeneratePageContentComplete";
+  VLOG(4) << "Contents(is_video=" << is_video
+          << ", invalidation_token=" << invalidation_token
+          << "): " << contents_text;
   if (navigation_id != current_navigation_id_) {
     VLOG(1) << __func__ << " for a different navigation. Ignoring.";
     return;
   }
 
   is_page_text_fetch_in_progress_ = false;
+
+  // If invalidation token matches existing token, then
+  // content was not re-fetched and we can use our existing cache.
+  if (!invalidation_token.empty() &&
+      (invalidation_token == content_invalidation_token_)) {
+    contents_text = article_text_;
+  } else {
+    is_video_ = is_video;
+    // Cache page content on instance so we don't always have to re-fetch
+    // if the content fetcher knows the content won't have changed and the fetch
+    // operation is expensive (e.g. network).
+    article_text_ = contents_text;
+    content_invalidation_token_ = invalidation_token;
+    engine_->SanitizeInput(article_text_);
+    // Update completion status
+    OnPageHasContentChanged(BuildSiteInfo());
+  }
+
+  on_page_text_fetch_complete_->Signal();
```

Comment:
DCHECK/CHECK `on_page_text_fetch_complete_` before signaling?

Reply:
How could it not be instantiated? It's first instantiated in the constructor.